### PR TITLE
Feature/template

### LINKED ISSUE #1: Template project for aquality-winappdriver-dotnet

### PR DESCRIPTION
**Pull request by paveliam**

### CODE REVIEW

**Comment by paveliam** [comment]
resolves #1

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        private static readonly IDictionary<Operation, string> OperationsLocators = new Dictionary<Operation, string>
+        {
+            { Operation.Add, "+" },
+            { Operation.Equals, "=" }
+        };
+
+        private ILabel ResultsLabel => ElementFactory.GetLabel(By.XPath("//*[@AutomationId='48']"), "Results bar");
+        
+        private IButton AboutTheAuthorButton => ElementFactory.GetButton(By.XPath("//*[@AutomationId='30']"), "About the Author");
```
Comment (mialeska) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs]: xpath locators are not very productive in general. For AutomationId attributes it's better to use MobileBy.AccessibilityId("30")

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/AboutTheAuthorWindow.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Aquality.WinAppDriver.Template.Forms
+{
+    public class AboutTheAuthorWindow : Window
+    {
+        private readonly IButton CloseButton;
+
+        public AboutTheAuthorWindow() : base(By.Name("About the Author"), "About the Author")
+        {
+            CloseButton = ElementFactory.FindChildElement<IButton>(this, By.Name("CLOSE"), "Close");
```
Comment (mialeska) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/AboutTheAuthorWindow.cs]: we can just call FindChildElement method of the base class, without referencing of ElementFactory and without passing "this"

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        public CalculatorForm() : base(By.Name("Day Maxi Calc  v.1.5 Freeware"), "Calculator")
```
Comment (mialeska) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs]: just wonder if the version should be hardcoded in locator

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/StepDefinitions/CalculatorWindowSteps.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+namespace Aquality.WinAppDriver.Template.SpecFlow.StepDefinitions
+{
+    [Binding]
+    public sealed class CalculatorWindowSteps
+    {
+        private readonly CalculatorForm calculatorForm;
+
+        public CalculatorWindowSteps()
+        {
+            calculatorForm = new CalculatorForm();
```
Comment (mialeska) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/StepDefinitions/CalculatorWindowSteps.cs]: probably better to resolve it from BoDi, receiving as constructor parameter. Same could work  for other steps and windows

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/Features/Demo.feature [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿Feature: Demo
+
+@demo
+Scenario: Add two numbers
+	When I click 7 on the numpad
```
Comment (mialeska) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/Features/Demo.feature]: probably it's better to put parameters in single quotes, isn't it?

File: README.md [hunk 6]
```diff
@@ -1 +1,42 @@
...
+### Configuration
+[settings.json](https://github.com/aquality-automation/aquality-winappdriver-dotnet/blob/master/Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Resources/settings.json) file contains settings of Aquality WinAppDriver library. Additional information you can find [here](https://github.com/aquality-automation/aquality-winappdriver-dotnet/blob/master/README.md).
+
+[allureConfig.json](https://github.com/aquality-automation/aquality-winappdriver-dotnet/blob/master/Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/allureConfig.json) is a part of Allure Report configuration. See details [here](https://github.com/allure-framework/allure-csharp#configuration).
+
+**NOTE:**
+
+> Do not store application executable under source control as [here](https://github.com/aquality-automation/aquality-winappdriver-dotnet-template/tree/master/Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Resources/Applications).
+
+> This is done only for demo purpose. You have to implement task in your CI system that will download and install necessary version of your application under the test.
```
Comment (mialeska) [inline README.md]: not necessarily, e.g. I have a clickonce application which autoupdates on start. delivery of app builds/updates may vary

**Review by mialeska — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (paveliam) [reply README.md]: I don't argee. Tests and application should be separated. What if you want to run the same version of tests on different versions of application? Store all of them under source control?

Comment (mialeska) [reply README.md]: nope, never store them under source control. 
I'm talking about "You have to implement task in your CI system that will download and install necessary version of your application under the test." - not necessarily
I agree that tests and application should be separated, but delivery of app builds/updates may vary

**Review by mialeska — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (paveliam) [reply README.md]: Oh, I got it. Sure, it not necessary. Will update README

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/Hooks/ApplicationHooks.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        [BeforeScenario]
+        public void StartApplication()
+        {
+            if (!AqualityServices.IsApplicationStarted)
+            {
+                AqualityServices.Application.Launch();
+            }
+        }
+
+
```
Comment (knysh) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template.SpecFlow/Hooks/ApplicationHooks.cs]: redundant space

File: Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+{
+    public class CalculatorForm : Form
+    {
+        private static readonly IDictionary<Operation, string> OperationsLocators = new Dictionary<Operation, string>
+        {
+            { Operation.Add, "+" },
+            { Operation.Equals, "=" }
+        };
+
+        private ILabel ResultsLabel => ElementFactory.GetLabel(MobileBy.AccessibilityId("48"), "Results bar");
```
Comment (knysh) [inline Aquality.WinAppDriver.Template/Aquality.WinAppDriver.Template/Forms/CalculatorForm.cs:17]: in AboutTheAuthorWindow we initialize element in constructor, here as property, should we use one approach?

**Review by knysh — APPROVED** [review]

**Pull request merged**